# Exercises I

In [63]:
import numpy as np
import pandas as pd
import matplotlib as plt
import env
from pydataset import data

#### 1. Run python -m pip install pymysql from your terminal to install the mysql client (any folder is fine)
Done

#### 2. cd into your exercises folder for this module and run echo env.py >> .gitignore
Done

#### 3. Create a function named get_db_url. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.



In [14]:
def get_db_url(db_name):
    return (f'mysql+pymysql://{env.user}:{env.password}@{env.host}/{db_name}')
    

#### 4. Use your function to obtain a connection to the employees database.

In [21]:
url = get_db_url('employees')


In [19]:
pd.read_sql('''
    select * from employees
    limit 5;
    ''', get_db_url('employees'))

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

#### 5. Once you have successfully run a query:

####     a. Intentionally make a typo in the database url. What kind of error message do you see?



In [23]:
pd.read_sql('''
    select * from employees
    limit 5;
    ''', get_db_url('employes'))

OperationalError: (pymysql.err.OperationalError) (1044, "Access denied for user 'oneil_2100'@'%' to database 'employes'")
(Background on this error at: https://sqlalche.me/e/14/e3q8)

#### b. Intentionally make an error in your SQL query. What does the error message look like?


In [24]:
pd.read_sql('''
    select * from employes
    limit 5;
    ''', get_db_url('employees'))

ProgrammingError: (pymysql.err.ProgrammingError) (1146, "Table 'employees.employes' doesn't exist")
[SQL: 
    select * from employes
    limit 5;
    ]
(Background on this error at: https://sqlalche.me/e/14/f405)

#### 6. Read the employees and titles tables into two separate DataFrames.



In [44]:
employees = pd.read_sql('''
    select * from employees
    ;
    ''', get_db_url('employees'))
employees

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [43]:
titles = pd.read_sql('''
    select * from titles
    ;
    ''', get_db_url('employees'))
titles

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

#### 7. How many rows and columns do you have in each DataFrame? Is that what you expected?

employees: 443308 rows x 4 cols, titles: 300024 rows x 6 cols. Multiple titles to different employees. Expected.

In [46]:
employees.shape

(300024, 6)

#### 8. Display the summary statistics for each DataFrame.


In [38]:
employees.describe()

emp_no
count    100.000000
mean   10050.500000
std       29.011492
min    10001.000000
25%    10025.750000
50%    10050.500000
75%    10075.250000
max    10100.000000

In [39]:
titles.describe()

emp_no
count    100.000000
mean   10034.320000
std       18.656859
min    10001.000000
25%    10018.000000
50%    10035.500000
75%    10050.000000
max    10066.000000

#### 9. How many unique titles are in the titles DataFrame?



In [47]:
titles.title.unique()

array(['Senior Engineer', 'Staff', 'Engineer', 'Senior Staff',
       'Assistant Engineer', 'Technique Leader', 'Manager'], dtype=object)

In [48]:
titles.title.nunique()

7

#### 10. What is the oldest date in the to_date column?


In [49]:
titles.to_date.min()

datetime.date(1985, 3, 1)

In [50]:
titles.to_date.sort_values().head()

16064     1985-03-01
406358    1985-03-11
346216    1985-03-20
35527     1985-03-23
277603    1985-04-17
Name: to_date, dtype: object

#### 11. What is the most recent date in the to_date column?


In [51]:
titles.to_date.max()

datetime.date(9999, 1, 1)

In [52]:
titles.to_date[titles.to_date != titles.to_date.max()].max()

datetime.date(2002, 8, 1)

# Exercises II

#### 1. Copy the users and roles DataFrames from the examples above.


In [53]:
# Create the users DataFrame.

users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users


id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [54]:
# Create the roles DataFrame

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles


id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

#### 2. What is the result of using a right join on the DataFrames?



In [55]:
users.merge(roles, left_on='role_id', right_on='id', how='right', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0     1      admin        both
1   2.0    joe      2.0     2     author        both
2   3.0  sally      3.0     3   reviewer        both
3   4.0   adam      3.0     3   reviewer        both
4   NaN    NaN      NaN     4  commenter  right_only

#### 3. What is the result of using an outer join on the DataFrames?


In [56]:
users.merge(roles, left_on='role_id', right_on='id', how='outer', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0   1.0      admin        both
1   2.0    joe      2.0   2.0     author        both
2   3.0  sally      3.0   3.0   reviewer        both
3   4.0   adam      3.0   3.0   reviewer        both
4   5.0   jane      NaN   NaN        NaN   left_only
5   6.0   mike      NaN   NaN        NaN   left_only
6   NaN    NaN      NaN   4.0  commenter  right_only

#### 4. What happens if you drop the foreign keys from the DataFrames and try to merge them?





In [58]:
users.merge(roles, how='right', indicator=True)

id       name  role_id      _merge
0   1      admin      NaN  right_only
1   2     author      NaN  right_only
2   3   reviewer      NaN  right_only
3   4  commenter      NaN  right_only

#### 5. Load the mpg dataset from PyDataset.



In [65]:
mpg = data('mpg')
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact

In [69]:
type(mpg)

pandas.core.frame.DataFrame

#### 6. Output and read the documentation for the mpg dataset.



In [67]:
data('mpg', show_doc=True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




#### 7. How many rows and columns are in the dataset?
234 rows and 11 columns


In [70]:
mpg.shape

(234, 11)

#### 8. Check out your column names and perform any cleanup you may want on them.



In [72]:
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact

In [77]:
mpg.columns = ['manufacturer', 'model', 'displacement', 'year',
              'cylinders', 'transmission', 'drive_wheels',
               'city_mileage', 'hwy_mileage', 'fl', 'class']

#### 9. Display the summary statistics for the dataset.



In [78]:
mpg.describe()

displacement         year   cylinders  city_mileage  hwy_mileage
count    234.000000   234.000000  234.000000    234.000000   234.000000
mean       3.471795  2003.500000    5.888889     16.858974    23.440171
std        1.291959     4.509646    1.611534      4.255946     5.954643
min        1.600000  1999.000000    4.000000      9.000000    12.000000
25%        2.400000  1999.000000    4.000000     14.000000    18.000000
50%        3.300000  2003.500000    6.000000     17.000000    24.000000
75%        4.600000  2008.000000    8.000000     19.000000    27.000000
max        7.000000  2008.000000    8.000000     35.000000    44.000000

#### 10. How many different manufacturers are there?



In [113]:
len(mpg.manufacturer.value_counts())

15

In [115]:
mpg.manufacturer.unique()

array(['audi', 'chevrolet', 'dodge', 'ford', 'honda', 'hyundai', 'jeep',
       'land rover', 'lincoln', 'mercury', 'nissan', 'pontiac', 'subaru',
       'toyota', 'volkswagen'], dtype=object)

In [116]:
mpg.manufacturer.nunique()

15

#### 11. How many different models are there?



In [90]:
len(mpg.model.value_counts())

38

#### 12. Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.



In [93]:
mpg['mileage_difference'] = mpg.hwy_mileage - mpg.city_mileage
mpg

manufacturer   model  displacement  year  cylinders transmission  \
1           audi      a4           1.8  1999          4     auto(l5)   
2           audi      a4           1.8  1999          4   manual(m5)   
3           audi      a4           2.0  2008          4   manual(m6)   
4           audi      a4           2.0  2008          4     auto(av)   
5           audi      a4           2.8  1999          6     auto(l5)   
..           ...     ...           ...   ...        ...          ...   
230   volkswagen  passat           2.0  2008          4     auto(s6)   
231   volkswagen  passat           2.0  2008          4   manual(m6)   
232   volkswagen  passat           2.8  1999          6     auto(l5)   
233   volkswagen  passat           2.8  1999          6   manual(m5)   
234   volkswagen  passat           3.6  2008          6     auto(s6)   

    drive_wheels  city_mileage  hwy_mileage fl    class  mileage_difference  
1              f            18           29  p  compact                  11  
2              f            21           29  p  compact                   8  
3              f            20           31  p  compact                  11  
4              f            21           30  p  compact                   9  
5              f            16           26  p  compact                  10  
..           ...           ...          ... ..      ...                 ...  
230            f            19           28  p  midsize                   9  
231            f            21           29  p  midsize                   8  
232            f            16           26  p  midsize                  10  
233            f            18           26  p  midsize                   8  
234            f            17           26  p  midsize                   9  

[234 rows x 12 columns]

#### 13. Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.




In [117]:
mpg['average_mileage2'] = mpg[['hwy_mileage', 'city_mileage']].agg('mean',axis=1)
mpg

manufacturer   model  displacement  year  cylinders transmission  \
1           audi      a4           1.8  1999          4     auto(l5)   
2           audi      a4           1.8  1999          4   manual(m5)   
3           audi      a4           2.0  2008          4   manual(m6)   
4           audi      a4           2.0  2008          4     auto(av)   
5           audi      a4           2.8  1999          6     auto(l5)   
..           ...     ...           ...   ...        ...          ...   
230   volkswagen  passat           2.0  2008          4     auto(s6)   
231   volkswagen  passat           2.0  2008          4   manual(m6)   
232   volkswagen  passat           2.8  1999          6     auto(l5)   
233   volkswagen  passat           2.8  1999          6   manual(m5)   
234   volkswagen  passat           3.6  2008          6     auto(s6)   

    drive_wheels  city_mileage  hwy_mileage fl    class  mileage_difference  \
1              f            18           29  p  compact                  11   
2              f            21           29  p  compact                   8   
3              f            20           31  p  compact                  11   
4              f            21           30  p  compact                   9   
5              f            16           26  p  compact                  10   
..           ...           ...          ... ..      ...                 ...   
230            f            19           28  p  midsize                   9   
231            f            21           29  p  midsize                   8   
232            f            16           26  p  midsize                  10   
233            f            18           26  p  midsize                   8   
234            f            17           26  p  midsize                   9   

     average_mileage  is_automatic  average_mileage2  
1               23.5          True              23.5  
2               25.0         False              25.0  
3               25.5         False              25.5  
4               25.5          True              25.5  
5               21.0          True              21.0  
..               ...           ...               ...  
230             23.5          True              23.5  
231             25.0         False              25.0  
232             21.0          True              21.0  
233             22.0         False              22.0  
234             21.5          True              21.5  

[234 rows x 15 columns]

In [96]:
mpg['average_mileage'] = mpg[['hwy_mileage', 'city_mileage']].mean(axis=1)
mpg

manufacturer   model  displacement  year  cylinders transmission  \
1           audi      a4           1.8  1999          4     auto(l5)   
2           audi      a4           1.8  1999          4   manual(m5)   
3           audi      a4           2.0  2008          4   manual(m6)   
4           audi      a4           2.0  2008          4     auto(av)   
5           audi      a4           2.8  1999          6     auto(l5)   
..           ...     ...           ...   ...        ...          ...   
230   volkswagen  passat           2.0  2008          4     auto(s6)   
231   volkswagen  passat           2.0  2008          4   manual(m6)   
232   volkswagen  passat           2.8  1999          6     auto(l5)   
233   volkswagen  passat           2.8  1999          6   manual(m5)   
234   volkswagen  passat           3.6  2008          6     auto(s6)   

    drive_wheels  city_mileage  hwy_mileage fl    class  mileage_difference  \
1              f            18           29  p  compact                  11   
2              f            21           29  p  compact                   8   
3              f            20           31  p  compact                  11   
4              f            21           30  p  compact                   9   
5              f            16           26  p  compact                  10   
..           ...           ...          ... ..      ...                 ...   
230            f            19           28  p  midsize                   9   
231            f            21           29  p  midsize                   8   
232            f            16           26  p  midsize                  10   
233            f            18           26  p  midsize                   8   
234            f            17           26  p  midsize                   9   

     average_mileage  
1               23.5  
2               25.0  
3               25.5  
4               25.5  
5               21.0  
..               ...  
230             23.5  
231             25.0  
232             21.0  
233             22.0  
234             21.5  

[234 rows x 13 columns]

#### 14. Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission.




In [ ]:
# mpg['is_automatic'] = mpg.transmission.apply(lambda x: 'auto' in x)

In [ ]:
# mpg['is_automatic'] = mpg.transmission.str.contains('auto')

In [102]:
mpg['is_automatic'] = mpg.transmission.str.startswith('auto')
mpg

manufacturer   model  displacement  year  cylinders transmission  \
1           audi      a4           1.8  1999          4     auto(l5)   
2           audi      a4           1.8  1999          4   manual(m5)   
3           audi      a4           2.0  2008          4   manual(m6)   
4           audi      a4           2.0  2008          4     auto(av)   
5           audi      a4           2.8  1999          6     auto(l5)   
..           ...     ...           ...   ...        ...          ...   
230   volkswagen  passat           2.0  2008          4     auto(s6)   
231   volkswagen  passat           2.0  2008          4   manual(m6)   
232   volkswagen  passat           2.8  1999          6     auto(l5)   
233   volkswagen  passat           2.8  1999          6   manual(m5)   
234   volkswagen  passat           3.6  2008          6     auto(s6)   

    drive_wheels  city_mileage  hwy_mileage fl    class  mileage_difference  \
1              f            18           29  p  compact                  11   
2              f            21           29  p  compact                   8   
3              f            20           31  p  compact                  11   
4              f            21           30  p  compact                   9   
5              f            16           26  p  compact                  10   
..           ...           ...          ... ..      ...                 ...   
230            f            19           28  p  midsize                   9   
231            f            21           29  p  midsize                   8   
232            f            16           26  p  midsize                  10   
233            f            18           26  p  midsize                   8   
234            f            17           26  p  midsize                   9   

     average_mileage  is_automatic  
1               23.5          True  
2               25.0         False  
3               25.5         False  
4               25.5          True  
5               21.0          True  
..               ...           ...  
230             23.5          True  
231             25.0         False  
232             21.0          True  
233             22.0         False  
234             21.5          True  

[234 rows x 14 columns]

#### 15. Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average?



In [118]:
mpg.groupby('manufacturer').average_mileage.mean().nlargest(1)

manufacturer
honda    28.5
Name: average_mileage, dtype: float64

#### 16. Do automatic or manual cars have better miles per gallon?



In [112]:
mpg.groupby('is_automatic').average_mileage.mean()

is_automatic
False    22.227273
True     19.130573
Name: average_mileage, dtype: float64